Tests/MathLib/mechanics_invariants.cpp

```diff
@@ -0,0 +1,286 @@
+        f, 1000,
+        ac::make_arbitrary(ac::map(&setZcomponentZero, kelvinVectorGenerator)),
+        gtest_reporter);
+}
+
+TEST_F(MathLibMechanicsInvariants6, SelfTestMappingKelvinToTensor)
+{
+    auto f = [](KelvinVector<6> const& v) {
+        return (v - tensorToKelvin<6>(kelvinToTensor(v))).norm() <=
+               1.5 * std::numeric_limits<double>::epsilon() * v.norm();
+    };
+
+    ac::check<KelvinVector<6>>(
+        f, 1000, ac::make_arbitrary(kelvinVectorGenerator), gtest_reporter);
+}
+
+//
+// Determinants of a Kelvin vector and corresponding Eigen::Matrix are equal.
+//
+
+TEST_F(MathLibMechanicsInvariants1, Determinant)
+{
+    auto f = [](KelvinVector<1> const& v) {
+        return Mechanics::Invariants<1>::determinant(v) ==
+               kelvinToTensor(v).determinant();
+    };
+
+    ac::check<KelvinVector<1>>(
+        f, 1000, ac::make_arbitrary(kelvinVectorGenerator), gtest_reporter);
+}
+
+TEST_F(MathLibMechanicsInvariants4, Determinant)
+{
+    auto f = [](KelvinVector<4> const& v) {
+        return std::abs(Mechanics::Invariants<4>::determinant(v) -
+                        kelvinToTensor(v).determinant()) <=
+               1e-12 * std::pow(v.norm(), 4);
+    };
+
+    ac::check<KelvinVector<4>>(
+        f, 1000,
+        ac::make_arbitrary(ac::map(&setZcomponentZero, kelvinVectorGenerator)),
+        gtest_reporter);
+}
+
+TEST_F(MathLibMechanicsInvariants6, Determinant)
+{
+    auto f = [](KelvinVector<6> const& v) {
+        return std::abs(Mechanics::Invariants<6>::determinant(v) -
+                        kelvinToTensor(v).determinant()) <=
+               std::numeric_limits<double>::epsilon() *
+                   std::pow(v.norm(), 3.07);
+    };
+
+    ac::check<KelvinVector<6>>(
+        f, 1000, ac::make_arbitrary(kelvinVectorGenerator), gtest_reporter);
+}
+
+//
+// Inverse of a Kelvin vector and corresponding Eigen::Matrix are equal.
+//
+
+TEST_F(MathLibMechanicsInvariants1, Inverse)
+{
+    auto f = [](KelvinVector<1> const& v) {
+        return Mechanics::Invariants<1>::inverse(v) ==
+               tensorToKelvin<1>(kelvinToTensor(v).inverse());
+    };
+
+    auto eps = std::numeric_limits<double>::epsilon();
+    ac::check<KelvinVector<1>>(
+        f, 1000, ac::make_arbitrary(kelvinVectorGenerator)
+                     .discard_if([&eps](KelvinVector<1> const& v) {
+                         return (v[0] < eps);  // only invertable matrices
+                     }),
+        gtest_reporter);
+}
+
+TEST_F(MathLibMechanicsInvariants4, Inverse)
+{
+    auto f = [](KelvinVector<4> const& v) {
+        return (Mechanics::Invariants<4>::inverse(v) -
+                tensorToKelvin<4>(kelvinToTensor(v).inverse()))
+                   .norm() < 1e-10 * std::pow(v.norm(), 2.7);
+    };
+
+    auto eps = std::numeric_limits<double>::epsilon();
+    ac::check<KelvinVector<4>>(
+        f, 1000,
+        ac::make_arbitrary(ac::map(&setZcomponentZero, kelvinVectorGenerator))
```

Comment:
If you set the z component to zero, isn't then the determinant of the Kelvin vector always zero, because then the corresponding 3x3 tensor should be
```
/ a b 0 \
| b c 0 |
\ 0 0 d /
```
with `d == 0`. Did I get something wrong?